### PR TITLE
Mrs Peacock in The Library with The Candlestick (killed main.{h,cpp})

### PR DESCRIPTION
**Pull request by TheBlueMatt**
🎵I can show you the world
Shining, shimmering splendid
Tell me, princess, now when did
You last let your heart decide!

I can open your eyes
Take you wonder by wonder
Over sideways and under
On a magic carpet ride

A whole new world
A new fantastic point of view
No one to tell us no
Or where to go
Or say we're only dreaming

A whole new world
A dazzling place I never knew
But when I'm way up here
It's crystal clear
That now I'm in a whole new world with you

Now I'm in a whole new world with you

Unbelievable sights
Indescribable feeling
Soaring, tumbling, freewheeling
Through an endless diamond sky
A whole new world
(Don't you dare close your eyes)
A hundred thousand things to see
(Hold your breath it gets better)
I'm like a shooting star
I've come so far
I can't go back to where I used to be

A whole new world
(Every turn a surprise)
With new horizons to pursue
(Every moment, red-letter)

I'll chase them anywhere
There's time to spare
Let me share this whole new world with you

A whole new world
(A whole new world)
That's where we'll be
(That's where we'll be)

A thrilling chase

A wondrous place

For you and me🎵

Lyrics Copyright © Warner/Chappell Music, Inc, Walt Disney Music Company, Universal Music Publishing Group

### CODE REVIEW

**Comment by TheBlueMatt** [comment]
Note that the main.h move is in its own commit separate from the main.h wrapper-file creation so that git tracks the rename and (at least some) rebases will be possible.

**Comment by sipa** [comment]
Verified 3b92e2d52fab7f5fccb5fd5e09cf31ee8327db32 to be move-only, apart from some header includes and:
```diff
-class CMainCleanup
+class CNetProcessingCleanup
 {
 public:
-    CMainCleanup() {}
-    ~CMainCleanup() {
-        // block headers
-        BlockMap::iterator it1 = mapBlockIndex.begin();
-        for (; it1 != mapBlockIndex.end(); it1++)
-            delete (*it1).second;
-        mapBlockIndex.clear();
-
+    CNetProcessingCleanup() {}
+    ~CNetProcessingCleanup() {
         // orphan transactions
         mapOrphanTransactions.clear();
         mapOrphanTransactionsByPrev.clear();
     }
-} instance_of_cmaincleanup;
+} instance_of_cnetprocessingcleanup;
```

**Comment by TheBlueMatt** [comment]
Nevermind, git apparently doesnt track across this kind of move if they are both in a merge commit, so instead ran sed s/#include "main.h"/#include "validation.h"/

**Comment by TheBlueMatt** [comment]
Fixed includes, diff from 3b92e2d52fab7f5fccb5fd5e09cf31ee8327db32 to 84922e4bf4c38227fbbbede50e09c87fe2a5c7f0 is

```
diff --git a/src/validation.cpp b/src/validation.cpp
index 08902bf..9cfb522 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -6,7 +6,6 @@
 #include "validation.h"
 
 #include "arith_uint256.h"
-#include "blockencodings.h"
 #include "chainparams.h"
 #include "checkpoints.h"
 #include "checkqueue.h"
@@ -15,7 +14,6 @@
 #include "consensus/validation.h"
 #include "hash.h"
 #include "init.h"
-#include "merkleblock.h"
 #include "policy/fees.h"
 #include "policy/policy.h"
 #include "pow.h"
diff --git a/src/zmq/zmqpublishnotifier.cpp b/src/zmq/zmqpublishnotifier.cpp
index 6a778da..99d42ab 100644
--- a/src/zmq/zmqpublishnotifier.cpp
+++ b/src/zmq/zmqpublishnotifier.cpp
@@ -3,6 +3,7 @@
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
 #include "chainparams.h"
+#include "streams.h"
 #include "zmqpublishnotifier.h"
 #include "validation.h"
 #include "util.h"
```

**Comment by sipa** [comment]
utACK d90cfde1c5edada8b524ee563c3daf0b71230fcf

Verified that 84922e4bf4c38227fbbbede50e09c87fe2a5c7f0 is move-only: only header includes changes and
```diff
-class CMainCleanup
+class CNetProcessingCleanup
 {
 public:
-    CMainCleanup() {}
-    ~CMainCleanup() {
-        // block headers
-        BlockMap::iterator it1 = mapBlockIndex.begin();
-        for (; it1 != mapBlockIndex.end(); it1++)
-            delete (*it1).second;
-        mapBlockIndex.clear();
-
+    CNetProcessingCleanup() {}
+    ~CNetProcessingCleanup() {
         // orphan transactions
         mapOrphanTransactions.clear();
         mapOrphanTransactionsByPrev.clear();
     }
-} instance_of_cmaincleanup;
+} instance_of_cnetprocessingcleanup;
```

**Comment by jtimon** [comment]
Fast review ACK 585a190 , but I didn't verified moveonly-ness, I didn't even tried to compile and I didn't read the full description of the PR.

**Comment by jonasschnelli** [comment]
utACK 585a190c97e7e20b35bc86f3f677c9f0b39e6dc9.
Testing this now on current master (= 6a2379cfabe27574d8d63010f5af38b98ee7fc4d).

File: src/net_processing.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+bool ProcessMessages(CNode* pfrom, CConnman& connman);
+/**
+ * Send queued protocol messages to be sent to a give node.
+ *
+ * @param[in]   pto             The node which we are sending messages to.
+ * @param[in]   connman         The connection manager for that node.
+ */
+bool SendMessages(CNode* pto, CConnman& connman);
+
+#endif // BITCOIN_MAIN_H
```
Comment (paveljanik) [inline src/net_processing.h]: wrong comment.

File: src/net_processing.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
+// Copyright (c) 2009-2010 Satoshi Nakamoto
+// Copyright (c) 2009-2015 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_NET_PROCESSING_H
```
Comment (paveljanik) [inline src/net_processing.h]: this header include in `.cpp` file?

File: src/net_processing.h (hunk 1 above)
Comment (paveljanik) [inline src/net_processing.h]: wrong comment

**Comment by instagibbs** [comment]
Might be meaningless but should cs_main be renamed as well?

File: src/net_processing.h (hunk 2 above)
Comment (paveljanik) [inline src/net_processing.h]: Hmm, looks like I reviewed it without clicking on "Load diff" and this puzzled me. Can be ignored.

**Comment by TheBlueMatt** [comment]
Fixed wrong comment in net_processing.h, squashed.
@instagibbs I'm ok with sed'ing the headers of most files, but cs_main maybe should come later

**Comment by morcos** [comment]
lightly tested ACK 76faa3cdfedbd3fc

**Review by sdaftuar — APPROVED** [review]
Couple nits, but not blockers to merging.

ACK 76faa3cdfedbd3fc91be4ecfff77fc6dc18134fb

File: src/net_processing.cpp [hunk 3]
```diff
@@ -0,0 +1,3026 @@
...
+#include "util.h"
+#include "utilmoneystr.h"
+#include "utilstrencodings.h"
+#include "validationinterface.h"
+
+#include <boost/thread.hpp>
+
+using namespace std;
+
+#if defined(NDEBUG)
```
Comment (TheBlueMatt) [inline src/net_processing.cpp]: Yea, I figured I'd just leave it in both.

File: src/main.cpp [hunk 4]
```diff
@@ -4311,8 +4311,6 @@ void UnloadBlockIndex()
     pindexBestInvalid = NULL;
     pindexBestHeader = NULL;
     mempool.clear();
-    mapOrphanTransactions.clear();
-    mapOrphanTransactionsByPrev.clear();
```
Comment (TheBlueMatt) [inline src/main.cpp]: I'd kinda prefer to leave it, actually...adding more extern ... mapOrphanTransactions to more files seems super gross, and we already have a check that mapOrphanTransactions{ByPrev} is empty at the end of all tests that use it, so I think its fine.

**Comment by sdaftuar** [comment]
Fine with me to not address in this pr.

**Comment by theuni** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/9260/commits/76faa3cdfedbd3fc91be4ecfff77fc6dc18134fb.

My audit process:
1. Verify the first commit is just a rename (and oneliner header fixup)
    - ```git diff HEAD~1:./main.cpp HEAD:./validation.cpp```
2. Check net_processing.h/cpp against what was removed from main:
    - ```git diff --patience HEAD~2..HEAD~1 -- main.cpp | grep '^-' | sed "s/^-//" > net_processing.cpp``` 
    - ```git diff --patience HEAD~2..HEAD~1 -- main.h | grep '^-' | sed "s/^-//" > net_processing.h```
    - ```git diff net_processing.cpp net_processing.h```
3. make sure nothing sneaky was added to main (now validation.h/cpp)
    - ```git diff -U0 --patience HEAD~2..HEAD~1 -- main.cpp main.h | grep -v '^[-@]'```
4. Checked for any static init issues between the newly split files

**Comment by sipa** [comment]
utACK 76faa3cdfedbd3fc91be4ecfff77fc6dc18134fb

My review strategy:
* Verify the the changes in the first and third commit by hand (they're all trivial).
* Use `git diff --patience e736772c~:src/main.h e736772c:src/main.h`, `git diff --patience e736772c~:src/main.h e736772c:src/net_processing.h` to determine that the new main.h and net_processing.h only contain code moved or copied from the old main.h.
* Use `git diff --patience e736772c~:src/main.cpp e736772c:src/main.cpp`, `git diff --patience e736772c~:src/main.cpp e736772c:src/net_processing.cpp` to determine that the new main.cpp and net_processing.cpp only contain code moved or copied from the old main.cpp.

The last item results in a few differences:
* Header include changes
* ```diff
             if (timeNow > pto->nextSendTimeFeeFilter) {
+                static FeeFilterRounder filterRounder(::minRelayTxFee);
                 CAmount filterToSend = filterRounder.round(currentFilter);
```
* ```diff
-class CMainCleanup
+class CNetProcessingCleanup
 {
 public:
-    CMainCleanup() {}
-    ~CMainCleanup() {
-        // block headers
-        BlockMap::iterator it1 = mapBlockIndex.begin();
-        for (; it1 != mapBlockIndex.end(); it1++)
-            delete (*it1).second;
-        mapBlockIndex.clear();
-
+    CNetProcessingCleanup() {}
+    ~CNetProcessingCleanup() {
         // orphan transactions
         mapOrphanTransactions.clear();
         mapOrphanTransactionsByPrev.clear();
     }
-} instance_of_cmaincleanup;
+} instance_of_cnetprocessingcleanup;
```

**Comment by rebroad** [comment]
Ok, so I have over 100 commits and branches that involve main.cpp. Is there a way I can rebase them all without needed to rewrite them all from scratch please?

**Comment by TheBlueMatt** [comment]
Commits to the validation side should auto-rebase, but commits to the net_processing side you'll have to do. If the changes are only to the net_processing side you could dump the diff and edit which file it is applying to.

**Comment by skang404** [comment]
Could someone comment as to why this renaming was done?

**Comment by jonasschnelli** [comment]
@skang404: The pull request description is indeed a little bit misleading 😊. `main.cpp` was getting a central container for all sorts of functions and logic. People just added stuff there instead of nicely layer and split stuff into separate files or classes.

This PR finally "removes" `main.cpp` (I guess if you have a `main.cpp` is a project large as this one you should consider refactoring some stuff) and splits the functions into a validation and net-processing part.

This was long overdue.

**Comment by jtimon** [comment]
I think @skang404 more specifically refers to renaming main.o to validation.o after separating the network processing code. I really don't see the point either given that's not even moved to another directory.
I complained on IRC but didn't want to slow down the separation of net_processing.o.

**Comment by TheBlueMatt** [comment]
Why not? git magically tracks that rename across rebase/merge, and the stuff there clearly fits into one category after the split.

**Comment by jtimon** [comment]
> Why not?

Right, the only disruption comes from includes. But I think the question is "why yes?".
In any case, I think it is too late to revert it (that would mean the same include disruption again in the other direction), so I don't think there's much point discussing this.

**Pull request merged**